File: splearn/src/main/java/com/splearn/domain/Member.java
package com.splearn.domain;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.ToString;

import static java.util.Objects.requireNonNull;
import static lombok.AccessLevel.PROTECTED;
import static org.springframework.util.Assert.state;

@Entity
@Getter
@ToString
@NoArgsConstructor(access = PROTECTED)
public class Member {

    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Embedded
    private Email email;

    private String nickname;

    private String passwordHash;

    @Enumerated(EnumType.STRING)
    private MemberStatus status;

    public static Member register(MemberRegisterRequest registerRequest, PasswordEncoder passwordEncoder) {
        Member member = new Member();

        member.email = new Email(requireNonNull(registerRequest.email()));
        member.nickname = requireNonNull(registerRequest.nickname());
        member.passwordHash = requireNonNull(passwordEncoder.encode(registerRequest.password()));

        member.status = MemberStatus.PENDING;

        return member;
    }

    public void activate() {
        state(status == MemberStatus.PENDING, "Member is not in PENDING status");

        this.status = MemberStatus.ACTIVE;
    }

    public void deactivate() {
        state(status == MemberStatus.ACTIVE, "Member is not in ACTIVE status");

        this.status = MemberStatus.DEACTIVATED;
    }

    public boolean verifyPassword(String password, PasswordEncoder passwordEncoder) {
        return passwordEncoder.matches(password, this.passwordHash);
    }

    public void changeNickname(String nickname) {
        this.nickname = requireNonNull(nickname);
    }

    public void changePassword(String password, PasswordEncoder passwordEncoder) {
        this.passwordHash = passwordEncoder.encode(requireNonNull(password));
    }

    public boolean isActive() {
        return this.status == MemberStatus.ACTIVE;
    }
}

File: splearn/src/main/kotlin/cham/splearn/domain/Member.kt
package cham.splearn.domain

import cham.splearn.domain.MemberStatus.*
import jakarta.persistence.Embedded
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.Id

@Entity
@ConsistentCopyVisibility
data class Member private constructor(

    @Id @GeneratedValue
    val id: Long = 0,

    @Embedded
    val email: Email,

    var nickname: String,

    var passwordHash: String,

    @Enumerated(EnumType.STRING)
    var status: MemberStatus = PENDING
){

    protected constructor() : this(
        email = Email.of("default@example.com"),
        nickname = "",
        passwordHash = ""
    )

    companion object {
        fun register(createRequest: MemberRegisterRequest, passwordEncoder: PasswordEncoder): Member {
            require(createRequest.nickname.isNotBlank()) { "Nickname cannot be blank" }
            require(createRequest.password.isNotBlank()) { "Password cannot be blank" }

            return Member(
                email = createRequest.email,
                nickname = createRequest.nickname,
                passwordHash = passwordEncoder.encode(createRequest.password)
            )
        }
    }

    fun activate() {
        check(status == PENDING){ "Only members with PENDING status can be activated." }

        status = ACTIVE
    }

    fun isActive() = status == ACTIVE

    fun deactivate() {
        check(isActive()){ "Only members with ACTIVE status can be deactivated." }

        status = DEACTIVATED
    }

    fun verifyPassword(password: String, passwordEncoder: PasswordEncoder): Boolean {
        return passwordEncoder.matches(password, this.passwordHash)
    }

    fun changeNickname(nickname: String) {
        require(nickname.isNotBlank()) { "Nickname cannot be blank" }
        this.nickname = nickname
    }

    fun changePassword(password: String, passwordEncoder: PasswordEncoder) {
        require(password.isNotBlank()) { "Password cannot be blank" }
        this.passwordHash = passwordEncoder.encode(password)
    }


}